### PR TITLE
feat: Use `std::io::BufWriter` with a 64KB buffer for dumping data

### PR DESCRIPTION
**Pull request by larseggert**
Let's see if this makes things faster.

### CODE REVIEW

**Review by mxinden — COMMENTED** [review]
Missing a `<BufWriter as Write>::flush` on `fin`, right? Otherwise some data might remain in the buffer and measurements are hard to compare.

**Comment by larseggert** [comment]
Probably, although the docs say it is called (just if there is an error that remains unhandled.)

File: neqo-bin/src/bin/client/http3.rs [hunk 1]
```diff
@@ -300,8 +301,12 @@ impl StreamHandler for DownloadStreamHandler {
...
         }
 
-        if fin && self.out_file.is_none() {
-            println!("<FIN[{stream_id}]>");
+        if fin {
+            if self.buf_writer.is_none() {
+                println!("<FIN[{stream_id}]>");
+            } else {
+                self.buf_writer.take().unwrap().flush()?;
+            }
```
Comment (mxinden) [inline neqo-bin/src/bin/client/http3.rs:309]:
```suggestion
            if let Some(buf_writer) = self.buf_writer.take() {
                buf_writer.flush()?;
            } else {
                println!("<FIN[{stream_id}]>");
            }
```

**Review by mxinden — APPROVED** [review]

**Comment by larseggert** [comment]
That whole "write to file" functionality could be factored out.

File: neqo-bin/src/bin/client/http09.rs [hunk 2]
```diff
@@ -253,15 +253,21 @@ impl<'b> Handler<'b> {
                 return Ok(());
             }
             Some(maybe_out_file) => {
+                let mut buf_writer = maybe_out_file
+                    .take()
+                    .map(|file| BufWriter::with_capacity(BUFWRITER_BUFFER_SIZE, file));
```
Comment (mxinden) [inline neqo-bin/src/bin/client/http09.rs:258]: Why construct a new `BufWriter` on each `read` call? Shouldn't the `BufWriter` be wrapping the `File` for the lifetime of the `File`? That would be consistent with the `http3.rs` implementation.

File: neqo-bin/src/bin/client/http3.rs [hunk 3]
```diff
@@ -255,8 +255,10 @@ impl StreamHandlerType {
         match handler_type {
             Self::Download => {
                 let out_file = get_output_file(url, &args.output_dir, all_paths);
+                let buf_writer =
+                    out_file.map(|file| BufWriter::with_capacity(BUFWRITER_BUFFER_SIZE, file));
```
Comment (mxinden) [inline neqo-bin/src/bin/client/http3.rs:259]: How about moving this into `get_output_file`. `get_output_file` is called by both `http3` and `http09`, thus no need for duplication.

**Review by mxinden — COMMENTED** [review]
I combined the suggestions in https://github.com/larseggert/neqo/pull/19.

**Comment by larseggert** [comment]
Thanks for that, it's the better approach. Can you point the PR at the main repo?

**Comment by mxinden** [comment]
> Thanks for that, it's the better approach. Can you point the PR at the main repo?

:+1: see https://github.com/mozilla/neqo/pull/1756.

**Comment by larseggert** [comment]
Overtaken by #1756

**Pull request closed without merging**